### PR TITLE
feat: New Module `avm/res/container-instance/container-group`

### LINKED ISSUE #538: [Module Proposal]: `avm/res/container-instance/container-group` (bicep)
### Check for previous/existing GitHub issues/module proposals

- [X] I have checked for previous/existing GitHub issues/module proposals.

### Check this module doesn't already exist in the module indexes

- [X] I have checked for that this module doesn't already exist in the module indexes; or I'm proposing the module to be migrated from CARML/TFVM.

### Bicep or Terraform?

Bicep

### Module Classification?

Resource Module

### Module Name

avm/res/container-instance/container-group

### Module Details

Container Instance will be migrated from CARML

### Do you want to be the owner of this module?

Yes

### Module Owner's GitHub Username (handle)

JPEasier

### (Optional) Secondary Module Owner's GitHub Username (handle)

_No response_

### PR DESCRIPTION
**Pull request by JPEasier**
## Description

Migration of Microsoft.containerInstance/containerGroup from CARML to AVM
Closes https://github.com/Azure/Azure-Verified-Modules/issues/538 

| Pipelines |
|-----|
|[avm.res.container-instance.container-group](https://github.com/JPEasier/bicep-registry-modules/actions/workflows/avm.res.container-instance.container-group.yml) |
## Adding a new module

- [x] A proposal has been submitted and approved.
- [x] I have included "Closes #{module_proposal_issue_number}" in the PR description.
- [x] I have run `brm validate` locally to verify the module files.
- [x] I have run deployment tests locally to ensure the module is deployable.

### CODE REVIEW

File: .github/CODEOWNERS [hunk 1]
```diff
@@ -31,7 +31,7 @@
 /avm/res/compute/virtual-machine/ @Azure/avm-res-compute-virtualmachine-module-owners-bicep @Azure/avm-core-team-technical-bicep
 #/avm/res/compute/virtual-machine-scale-set/ @Azure/avm-res-compute-virtualmachinescaleset-module-owners-bicep @Azure/avm-core-team-technical-bicep
 /avm/res/consumption/budget/ @Azure/avm-res-consumption-budget-module-owners-bicep @Azure/avm-core-team-technical-bicep
-#/avm/res/container-instance/container-group/ @Azure/avm-res-containerinstance-containergroup-module-owners-bicep @Azure/avm-core-team-technical-bicep
+/avm/res/container-instance/container-group/ @Azure/avm-res-containerinstance-containergroup-module-owners-bicep @Azure/avm-core-team-technical-bicep
```
Comment (AlexanderSehr) [inline .github/CODEOWNERS:34]: Note, the team is either not yet existing or parent not yet approved

File: avm/res/container-instance/container-group/main.bicep [hunk 2]
```diff
@@ -0,0 +1,293 @@
...
+output resourceId string = containergroup.id
+
+@description('The resource group the container group was deployed into.')
+output resourceGroupName string = resourceGroup().name
+
+@description('The IPv4 address of the container group.')
+output iPv4Address string = containergroup.properties.ipAddress.ip
+
+@description('The principal ID of the system assigned identity.')
+output systemAssignedMIPrincipalId string = (managedIdentities.?systemAssigned ?? false) && contains(containergroup.identity, 'principalId') ? containergroup.identity.principalId : ''
```
Comment (AlexanderSehr) [inline avm/res/container-instance/container-group/main.bicep:187]:
```suggestion
output systemAssignedMIPrincipalId string = containergroup.?identity.?principalId ?? ''
```

File: avm/res/container-instance/container-group/main.bicep [hunk 3]
```diff
@@ -0,0 +1,293 @@
...
+      requests: {
+        cpu: int
+        gpu: {
+          count: int
+          sku: string
+        }?
+        memoryInGB: int
+      }
+
+      @description('Optional. The resource limits of this container instance..')
```
Comment (AlexanderSehr) [inline avm/res/container-instance/container-group/main.bicep:258]:
```suggestion
      @description('Optional. The resource limits of this container instance.')
```

File: avm/res/container-instance/container-group/main.bicep [hunk 4]
```diff
@@ -0,0 +1,293 @@
...
+    ports: {
+      @description('Required. The port number exposed on the container instance.')
+      port: int
+
+      @description('Required. The protocol associated with the port number.')
+      protocol: string
+    }[]
+
+    resources: {
+      @description('Required. The resource requests of this container instance..')
```
Comment (AlexanderSehr) [inline avm/res/container-instance/container-group/main.bicep:248]:
```suggestion
      @description('Required. The resource requests of this container instance.')
```

File: avm/res/container-instance/container-group/main.bicep [hunk 5]
```diff
@@ -0,0 +1,293 @@
...
+    }[]?
+
+    @description('Optional. The command to execute within the container instance.')
+    command: string[]?
+
+    @description('Optional. The environment variables to set in the container instance.')
+    environmentVariables: {
+      name: string
+      secureValue: string?
+      value: string?
```
Comment (AlexanderSehr) [inline avm/res/container-instance/container-group/main.bicep:282]: Would be good to have some metadata from the inner parameters to as those are rendered in the readme + are shown by the extension when using the UDTs. The same can be set about all inner properties in each UDT.

File: avm/res/container-instance/container-group/tests/e2e/encr/main.test.bicep [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+    customerManagedKey: {
+      keyName: nestedDependencies.outputs.keyVaultEncryptionKeyName
+      keyVaultResourceId: nestedDependencies.outputs.keyVaultResourceId
+      userAssignedIdentityResourceId: nestedDependencies.outputs.managedIdentityResourceId
+    }
+    tags: {
+      'hidden-title': 'This is visible in the resource name'
+      Environment: 'Non-Prod'
+      Role: 'DeploymentValidation'
+    }
```
Comment (AlexanderSehr) [inline avm/res/container-instance/container-group/tests/e2e/encr/main.test.bicep:134]:
```suggestion
```

Please remove any parameter/property that isn't required for CMK to make this test case 'as on point' as possible

File: avm/res/container-instance/container-group/tests/e2e/private/main.test.bicep [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+      systemAssigned: true
+      userAssignedResourceIds: [
+        nestedDependencies.outputs.managedIdentityResourceId
+      ]
+    }
+    tags: {
+      'hidden-title': 'This is visible in the resource name'
+      Environment: 'Non-Prod'
+      Role: 'DeploymentValidation'
+    }
```
Comment (AlexanderSehr) [inline avm/res/container-instance/container-group/tests/e2e/private/main.test.bicep:143]: Required for private networking?

File: avm/res/container-instance/container-group/tests/e2e/waf-aligned/main.test.bicep [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+        protocol: 'Tcp'
+        port: 443
+      }
+    ]
+    managedIdentities: {
+      systemAssigned: true
+      userAssignedResourceIds: [
+        nestedDependencies.outputs.managedIdentityResourceId
+      ]
+    }
```
Comment (AlexanderSehr) [inline avm/res/container-instance/container-group/tests/e2e/waf-aligned/main.test.bicep:119]: Required for WAF-alignment? If not, could be removed alongside the dependency

File: .github/CODEOWNERS (hunk 1 above)
Comment (ChrisSidebotham) [reply .github/CODEOWNERS:34]: @JPEasier - Please create the Github Team required for the module

Comment (JPEasier) [reply .github/CODEOWNERS:34]: https://github.com/orgs/Azure/teams/avm-res-containerinstance-containergroup-module-owners-bicep

Comment (JPEasier) [reply .github/CODEOWNERS:34]: https://github.com/orgs/Azure/teams/avm-res-containerinstance-containergroup-module-contributors-bicep

File: avm/res/container-instance/container-group/main.bicep (hunk 5 above)
Comment (JPEasier) [reply avm/res/container-instance/container-group/main.bicep]: put in description to each type param

File: avm/res/container-instance/container-group/main.bicep [hunk 9]
```diff
@@ -88,87 +88,87 @@ param customerManagedKey customerManagedKeyType
...
-      resources: []
-      outputs: {
-        telemetry: {
-          type: 'String'
-          value: 'For more information, see https://aka.ms/avm/TelemetryInfo'
-        }
-      }
-    }
-  }
+	name: '46d3xbcp.res.containerinstance-containergroup.${replace('-..--..-', '.', '-')}.${substring(uniqueString(deployment().name, location), 0, 4)}'
```
Comment (AlexanderSehr) [inline avm/res/container-instance/container-group/main.bicep:96]: @JPEasier beware - your indent settings seemd to have multiplied to an extreme. 
image
 Please undo.

Comment (JPEasier) [reply avm/res/container-instance/container-group/main.bicep:96]: Upsi...

File: avm/res/container-instance/container-group/tests/e2e/defaults/main.test.bicep [hunk 10]
```diff
@@ -36,37 +36,39 @@ resource resourceGroup 'Microsoft.Resources/resourceGroups@2021-04-01' = {
...
-              port: 443
-              protocol: 'Tcp'
-            }
-          ]
-          resources: {
-            requests: {
-              cpu: 2
-              memoryInGB: 2
+module testDeployment '../../../main.bicep' = [
+  for iteration in ['init', 'idem']: {
```
Comment (AlexanderSehr) [inline avm/res/container-instance/container-group/tests/e2e/defaults/main.test.bicep]: Question: Do you have the latest upstream version merged? I.e., does your readme generation still work?

Comment (JPEasier) [reply avm/res/container-instance/container-group/tests/e2e/defaults/main.test.bicep]: Yesterday I did the latest updates.

Comment (AlexanderSehr) [reply avm/res/container-instance/container-group/tests/e2e/defaults/main.test.bicep]: Now only the static tests fail. If you want to unblock this PR you'd need to remove the line break in the test files (only there, not the rest of the templates) - OR - wait for PR #1368

Comment (JPEasier) [reply avm/res/container-instance/container-group/tests/e2e/defaults/main.test.bicep]: Thanks @AlexanderSehr for the update 🙂

Comment (JPEasier) [reply avm/res/container-instance/container-group/tests/e2e/defaults/main.test.bicep]: @AlexanderSehr its done :)

**Review by AlexanderSehr — APPROVED** [review]

**Pull request merged**